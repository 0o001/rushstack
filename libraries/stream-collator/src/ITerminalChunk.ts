// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * This API was introduced as a temporary measure.
 * @beta
 */
export enum StreamKind {
  Stdout = 'O',
  Stderr = 'E'
}

/**
 * This API was introduced as a temporary measure.
 * @beta
 */
export interface ITerminalChunk {
  stream: StreamKind;
  text: string;
}
