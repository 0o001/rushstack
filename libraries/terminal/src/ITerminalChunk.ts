// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * This API was introduced as a temporary measure.
 * @beta
 */
export const enum TerminalChunkKind {
  Stdout = 'O',
  Stderr = 'E'
}

/**
 * This API was introduced as a temporary measure.
 * @beta
 */
export interface ITerminalChunk {
  kind: TerminalChunkKind;
  text: string;
}
