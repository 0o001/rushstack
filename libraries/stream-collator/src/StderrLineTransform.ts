// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Text, NewlineKind } from '@rushstack/node-core-library';

import { ITerminalChunk, StreamKind } from './ITerminalChunk';
import { TerminalTransform, ITerminalTransformOptions } from './TerminalTransform';

/** @beta */
export interface IStderrLineTransformOptions extends ITerminalTransformOptions {
  newlineKind?: NewlineKind;
}

/** @beta */
export class StderrLineTransform extends TerminalTransform {
  private _accumulatedLine: string;
  private _accumulatedStderr: boolean;

  public readonly newline: string;

  public constructor(options: IStderrLineTransformOptions) {
    super(options);

    this._accumulatedLine = '';
    this._accumulatedStderr = false;

    this.newline = Text.getNewline(options.newlineKind || NewlineKind.Lf);
  }

  protected onWriteChunk(chunk: ITerminalChunk): void {
    const text: string = Text.convertToLf(chunk.text);
    let startIndex: number = 0;

    while (startIndex < text.length) {
      if (chunk.stream === StreamKind.Stderr) {
        this._accumulatedStderr = true;
      }

      const endIndex: number = text.indexOf('\n', startIndex);
      if (endIndex < 0) {
        // we did not find \n, so simply append
        this._accumulatedLine += text.substring(startIndex);
        break;
      }

      // append everything up to \n
      this._accumulatedLine += text.substring(startIndex, endIndex);

      this._processAccumulatedLine();

      // skip the \n
      startIndex = endIndex + 1;
    }
  }

  protected onClose(): void {
    if (this._accumulatedLine.length > 0) {
      this._processAccumulatedLine();
    }
    this.destination.close();
  }

  private _processAccumulatedLine(): void {
    this._accumulatedLine += this.newline;

    if (this._accumulatedStderr) {
      this.destination.writeChunk({
        stream: StreamKind.Stderr,
        text: this._accumulatedLine
      });
    } else {
      this.destination.writeChunk({
        stream: StreamKind.Stdout,
        text: this._accumulatedLine
      });
    }

    this._accumulatedLine = '';
    this._accumulatedStderr = false;
  }
}
