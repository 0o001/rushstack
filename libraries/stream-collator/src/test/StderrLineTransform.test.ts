// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { StreamKind } from '../ITerminalChunk';
import { StderrLineTransform } from '../StderrLineTransform';
import { TestWriter } from '../TestWriter';

describe('LineAlignerStream', () => {
  it('should report stdout if there is no stderr', () => {
    const testWriter: TestWriter = new TestWriter();
    const transform: StderrLineTransform = new StderrLineTransform({ destination: testWriter });

    transform.writeChunk({ text: 'stdout 1\nstdout 2\n', stream: StreamKind.Stdout });
    transform.close();

    expect(testWriter.chunks).toMatchSnapshot();
  });
});
