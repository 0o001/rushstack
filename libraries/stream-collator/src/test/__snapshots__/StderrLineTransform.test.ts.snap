// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineAlignerStream should report stdout if there is no stderr 1`] = `
Array [
  Object {
    "kind": "O",
    "text": "stdout 1
",
  },
  Object {
    "kind": "O",
    "text": "stdout 2
",
  },
]
`;
